substrate/frame/revive/src/debug.rs

```diff
@@ -15,62 +15,155 @@
 
+use crate::{evm::TracerConfig, Config, DispatchError, Weight};
 pub use crate::{
+	evm::{CallLog, CallTrace, CallType, EthTraces, Traces},
 	exec::{ExecResult, ExportedFunction},
 	primitives::ExecReturnValue,
 };
-use crate::{Config, LOG_TARGET};
-use sp_core::H160;
+use alloc::{boxed::Box, format, string::ToString, vec::Vec};
+use sp_core::{H160, H256, U256};
 
 /// Umbrella trait for all interfaces that serves for debugging.
-pub trait Debugger<T: Config>: Tracing<T> + CallInterceptor<T> {}
+pub trait Debugger<T: Config>: CallInterceptor<T> {}
+impl<T: Config, V> Debugger<T> for V where V: CallInterceptor<T> {}
 
-impl<T: Config, V> Debugger<T> for V where V: Tracing<T> + CallInterceptor<T> {}
+/// Defines methods to capture contract calls
+pub trait Tracing {
+	/// Called before a contract call is executed
+	fn enter_child_span(
+		&mut self,
+		from: H160,
+		to: H160,
+		is_delegate_call: bool,
+		is_read_only: bool,
+		value: U256,
+		input: &[u8],
+		gas: Weight,
+	);
 
-/// Defines methods to capture contract calls, enabling external observers to
-/// measure, trace, and react to contract interactions.
-pub trait Tracing<T: Config> {
-	/// The type of [`CallSpan`] that is created by this trait.
-	type CallSpan: CallSpan;
+	/// Record a log event
+	fn log_event(&mut self, event: H160, topics: &[H256], data: &[u8]);
 
-	/// Creates a new call span to encompass the upcoming contract execution.
-	///
-	/// This method should be invoked just before the execution of a contract and
-	/// marks the beginning of a traceable span of execution.
-	///
-	/// # Arguments
-	///
-	/// * `contract_address` - The address of the contract that is about to be executed.
-	/// * `entry_point` - Describes whether the call is the constructor or a regular call.
-	/// * `input_data` - The raw input data of the call.
-	fn new_call_span(
-		contract_address: &H160,
-		entry_point: ExportedFunction,
-		input_data: &[u8],
-	) -> Self::CallSpan;
+	/// Called after a contract call is executed
+	fn exit_child_span(&mut self, output: &ExecReturnValue, gas_left: Weight);
+
+	/// Called when a contract call terminates with an error
+	fn exit_child_span_with_error(&mut self, error: DispatchError, gas_left: Weight);
+
+	/// Collects and returns the traces recorded by the tracer, then clears them.
+	fn collect_traces(&mut self) -> Traces;
 }
 
-/// Defines a span of execution for a contract call.
-pub trait CallSpan {
-	/// Called just after the execution of a contract.
-	///
-	/// # Arguments
-	///
-	/// * `output` - The raw output of the call.
-	fn after_call(self, output: &ExecReturnValue);
+/// Creates a new tracer from the given config.
+pub fn make_tracer(config: TracerConfig) -> Box<dyn Tracing> {
+	match config {
+		TracerConfig::CallTracer { with_logs } => Box::new(CallTracer::new(with_logs)),
+	}
 }
 
-impl<T: Config> Tracing<T> for () {
-	type CallSpan = ();
+#[derive(Default, Debug, Clone, PartialEq, Eq)]
+pub struct CallTracer {
```

Comment:
The concrete implementation belongs into the evm module. This module should not need to import anything from the evm module. I can imagine that there are alternative implementations provided when using pallet_revive as a debugger.